File: product-service/internal/services/currency_conversion_service.go
package services

import (
	"context"
	"fmt"
	"strings"

	"github.com/PedroScheurer/product-service/internal/clients"
	"github.com/PedroScheurer/product-service/internal/dtos"
	"github.com/PedroScheurer/product-service/internal/entities"
)

const (
	convertedValueCacheName = "ConvertedValue"
	fallbackPrice           = -1.0
)

// CurrencyConversionService é o equivalente Go da classe Java
// CurrencyConversionService: decide se precisa converter a moeda do
// produto, consultando o cache antes de chamar o currency-service via
// CurrencyClient, e atualizando o cache com o resultado.
type CurrencyConversionService struct {
	currencyClient clients.CurrencyClient
	cacheService   *CacheService
}

func NewCurrencyConversionService(currencyClient clients.CurrencyClient, cacheService *CacheService) *CurrencyConversionService {
	return &CurrencyConversionService{
		currencyClient: currencyClient,
		cacheService:   cacheService,
	}
}

// Convert é o equivalente a CurrencyConversionService.convert(product, targetCurrency, port).
func (s *CurrencyConversionService) Convert(ctx context.Context, product *entities.ProductEntity, targetCurrency, port string) dtos.ConversionResult {
	environment := "Product-service running on Port: " + port

	if isSameCurrency(product.Currency, targetCurrency) {
		return dtos.ConversionResult{
			ConvertedPrice: product.Price,
			Environment:    environment,
		}
	}

	keyCache := buildCacheKey(product.Currency, targetCurrency)

	if conversionRate, found := s.cacheService.Get(convertedValueCacheName, keyCache); found {
		return dtos.ConversionResult{
			ConvertedPrice: applyConversion(product.Price, conversionRate),
			Environment:    environment + " | Currency in cache",
		}
	}

	currency, err := s.currencyClient.GetCurrency(ctx, product.Currency, targetCurrency)
	if err != nil || currency == nil {
		return dtos.ConversionResult{
			ConvertedPrice: fallbackPrice,
			Environment:    environment + " | Currency fallback",
		}
	}

	s.cacheService.Put(convertedValueCacheName, keyCache, currency.ConversionRate)

	return dtos.ConversionResult{
		ConvertedPrice: applyConversion(product.Price, currency.ConversionRate),
		Environment:    fmt.Sprintf("%s | %s", environment, currency.Environment),
	}
}

func isSameCurrency(currency, targetCurrency string) bool {
	// Equivalente a targetCurrency.equalsIgnoreCase(currency) no Java.
	return strings.EqualFold(currency, targetCurrency)
}

func buildCacheKey(currency, targetCurrency string) string {
	return currency + " - " + targetCurrency
}

func applyConversion(price, conversionRate float64) float64 {
	return price * conversionRate
}

File: product-service/src/main/java/br/edu/atitus/productservice/services/CurrencyConversionService.java
package br.edu.atitus.productservice.services;

import br.edu.atitus.productservice.clients.CurrencyClient;
import br.edu.atitus.productservice.clients.CurrencyResponse;
import br.edu.atitus.productservice.dtos.ConversionResult;
import br.edu.atitus.productservice.entities.ProductEntity;
import org.springframework.stereotype.Service;

@Service
public class CurrencyConversionService {

    private static final String CACHE_NAME = "ConvertedValue";
    private static final Double FALLBACK_PRICE = -1.0;

    private final CurrencyClient currencyClient;
    private final CacheService cacheService;

    public CurrencyConversionService(CurrencyClient currencyClient, CacheService cacheService) {
        this.currencyClient = currencyClient;
        this.cacheService = cacheService;
    }

    public ConversionResult convert(ProductEntity product, String targetCurrency, String port) {
        String environment = "Product-service running on Port: " + port;

        if (isSameCurrency(product.getCurrency(), targetCurrency)) {
            return new ConversionResult(product.getPrice(), environment);
        }

        String keyCache = buildCacheKey(product.getCurrency(), targetCurrency);

        Double conversionRate = cacheService.get(CACHE_NAME, keyCache);

        if (conversionRate != null) {
            return new ConversionResult(
                    applyConversion(product.getPrice(), conversionRate),
                    environment + " | Currency in cache"
            );
        }

        CurrencyResponse currency = getCurrencyFromApi(product.getCurrency(), targetCurrency);
        System.out.println(currency);
        if (currency == null) {
            return new ConversionResult(
                    FALLBACK_PRICE,
                    environment + " | " + "Currency fallback"
            );
        }

        cacheService.put(CACHE_NAME, keyCache, currency.conversionRate());

        return new ConversionResult(
                applyConversion(product.getPrice(), currency.conversionRate()),
                environment + " | " + currency.environment()
        );
    }

    private CurrencyResponse getCurrencyFromApi(String currency, String targetCurrency) {
        return currencyClient.getCurrency(currency, targetCurrency);
    }

    private boolean isSameCurrency(String currency, String targetCurrency) {
        return targetCurrency.equalsIgnoreCase(currency);
    }

    private String buildCacheKey(String currency, String targetCurrency) {
        return currency + " - " + targetCurrency;
    }

    private Double applyConversion(Double price, Double conversionRate) {
        return price * conversionRate;
    }
}
